src/h5web/vis-packs/core/hooks.ts

```diff
@@ -16,11 +16,30 @@ export function useDatasetValue<
   S extends HDF5Shape,
   T extends HDF5Type,
   D extends Dataset<S, T> | undefined
->(dataset: D): D extends Dataset<infer S, infer T> ? Value<S, T> : undefined;
+>(
+  dataset: D,
+  dimMapping?: DimensionMapping,
+  fetchAll?: boolean
+): D extends Dataset<infer S, infer T> ? Value<S, T> : undefined;
+
+export function useDatasetValue(
+  dataset: Dataset | undefined,
+  dimMapping?: DimensionMapping,
+  fetchAll = false
+) {
+  /* Create slice selection string from dim mapping:
+   * - [0, 'y', 'x'] => "0,:,:"
+   * - [0, 'x'] with `fetchAll` => ":,:"
+   *
+   * Passing a `dimMapping` with `fetchAll` is equivalent to passing no `dimMapping` at all,
+   * but it solves some caching edge cases, like switching between HeatmapVis and LineVis
+   * with a 2D dataset, as it ensures the selection string is always ":,:". */
+  const selection = dimMapping
+    ?.map((dim) => (!fetchAll && isNumber(dim) ? dim : ':'))
+    .join(',');
 
-export function useDatasetValue(dataset: Dataset | undefined) {
   const { valuesStore } = useContext(ProviderContext);
-  return dataset && valuesStore.get(dataset.path);
+  return dataset && valuesStore.get({ path: dataset.path, selection });
 }
```

Comment:
I hope the comment is clear. I don't see a simpler way to explain why the `LineVis` has to call `useDatasetValue` with its `dimMapping` and the `fetchAll` flag set to `true`, and I prefer the comment to be here than in `LineVis`.